In [38]:
#import pandas, numpy, and os libraries
import pandas as pd
import numpy as np
import os
path = r'C:\Users\joelf\OneDrive\Documents\CareerFoundry\4 Python\Instacart Data'

In [39]:
#import now formated Orders dataset (see formating in 4.4)
df_ord = pd.read_csv(os.path.join(path, 'NewDataSets' ,'orders_wrangled.csv'), index_col = False)

#import Products
df_prods = pd.read_csv(os.path.join(path,'products.csv'), index_col = False)

4.5 Coursework of wrangling the products file and then the taskwork of wrangling orders file

In [40]:
#Finding and Removing Nulls
#count number of Nulls in data frame
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [41]:
#seeing shape of products before dropping nulls
df_prods.shape

(49693, 5)

In [42]:
#Given the low number of nulls and inability to fill them (since they are str), the Nulls will be dropped
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]
df_prods_clean.shape #making sure there are 16 less rows

(49677, 5)

In [43]:
#Finding duplicate rows in data
df_dups = df_prods_clean[df_prods_clean.duplicated()]
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [44]:
df_prods_clean.shape #to see how many rows before dups are dropped

(49677, 5)

In [45]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()
df_prods_clean_no_dups.shape #to make sure 5 rows were dropped

(49672, 5)

In [46]:
#Export as products_wrangles
df_prods_clean_no_dups.to_csv(os.path.join(path, 'NewDataSets' , 'products_wrangled.csv'))

4.5 Taskwork with Orders_wrangled Dataset

In [47]:
#1 above
#2
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

For the Products data the prices need to be looked into. Right now the largest price is set to 99,999.

In [48]:
#3 Check for mixed data types
for col in df_ord.columns.tolist():
	weird = (df_ord[[col]].applymap(type) != df_ord[[col]].iloc[0].apply(type)).any(axis = 1)
	if len (df_ord[weird]) > 0:
		print (col)


No mixed data types found in Orders

In [49]:
#5 Check for missing values
df_ord.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_time                     0
days_since_prior_order    206209
dtype: int64

Missing values only appear in days_since_prior_order. This could easily be explained by customers who dont have a prior order
ie they only ordered once

In [50]:
#6 Address missing values
#Create a new column with a flag
df_ord['repeat_customer'] = np.where(df_ord['days_since_prior_order'].isnull(), 0, 1)
df_ord.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_time  days_since_prior_order  repeat_customer  
0           8                     NaN                0  
1           7                    15.0                1  
2          12                    21.0                1  
3           7                    29.0                1  
4          15                    28.0                1

I created a flag that can be used to filter out the one time customers. The purchase data for one time customers is valid and important but if there is a calculation that throws an error based on days_since_prior_order having a Null, this filter can take care of that issue

In [51]:
#7 Check for Duplicates
df_ord_dups = df_ord[df_ord.duplicated()]
df_ord_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_time, days_since_prior_order, repeat_customer]
Index: []

#7 and #8 No duplicate values found in Orders Dataframe

In [52]:
#9 Export
df_ord.to_csv(os.path.join(path, 'NewDataSets', 'orders_wrangled.csv'))
#products_wrangled exported above in Coursework